discord/ext/commands/core.py

```diff
@@ -984,31 +1005,44 @@ def short_doc(self):
         return ''
 
     def _is_typing_optional(self, annotation):
-        try:
-            origin = annotation.__origin__
-        except AttributeError:
-            return False
+        return typing.get_origin(annotation) is typing.Union and typing.get_args(annotation)[-1] is type(None)
 
-        if origin is not typing.Union:
-            return False
+    def _is_typing_literal(self, annotation):
+        return typing.get_origin(annotation) is typing.Literal
 
-        return annotation.__args__[-1] is type(None)
+    def _flattened_typing_literal_args(self, annotation):
+        for literal in typing.get_args(annotation):
+            if self._is_typing_literal(literal):
+                yield from self._flattened_typing_literal_args(literal)
+            else:
+                yield literal
 
     @property
     def signature(self):
         """:class:`str`: Returns a POSIX-like signature useful for help command output."""
         if self.usage is not None:
             return self.usage
 
-
         params = self.clean_params
         if not params:
             return ''
 
+        def value_str(value):
+            return f'"{value}"' if isinstance(value, str) else str(value)
+
         result = []
         for name, param in params.items():
             greedy = isinstance(param.annotation, converters._Greedy)
 
+            # for typing.Literal[...], typing.Optional[typing.Literal[...]], and Greedy[typing.Literal[...]], the
+            # parameter signature is a literal list of its values
+            if self._is_typing_literal(param.annotation):
+                name = '|'.join(value_str(v) for v in self._flattened_typing_literal_args(param.annotation))
+            elif self._is_typing_optional(param.annotation) and self._is_typing_literal(typing.get_args(param.annotation)[0]):
+                name = '|'.join(value_str(v) for v in self._flattened_typing_literal_args(typing.get_args(param.annotation)[0]))
+            elif greedy and self._is_typing_literal(param.annotation.converter):
+                name = '|'.join(value_str(v) for v in self._flattened_typing_literal_args(param.annotation.converter))
```

Comment:
There's a lot of code duplication here and a lot of double checking. The performance of this could probably be better as a result.

Reply:
Should hopefully be more optimised now.